### DISCUSSION TITLE
Notes for dealing with deprecation warning in vrv.cpp in C++20

### DISCUSSION THREAD

**Discussion by craigsapp** [rism-digital/verovio#3651] [Ideas]
In vrv.cpp, there are deprecation warnings:

```
 Building CXX object CMakeFiles/verovio.dir/Users/craig/git-cloud/vhv/verovio/src/vrv.cpp.o
/Users/craig/git-cloud/vhv/verovio/src/vrv.cpp:267:31: warning: 'codecvt_utf8<char32_t, 1114111, 0>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> strCnv;
                              ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/codecvt:199:28: note: 'codecvt_utf8<char32_t, 1114111, 0>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 codecvt_utf8
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
/Users/craig/git-cloud/vhv/verovio/src/vrv.cpp:267:10: warning: 'wstring_convert<std::codecvt_utf8<char32_t, 1114111, 0>, char32_t>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> strCnv;
         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/locale:3621:28: note: 'wstring_convert<std::codecvt_utf8<char32_t, 1114111, 0>, char32_t>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 wstring_convert
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
/Users/craig/git-cloud/vhv/verovio/src/vrv.cpp:273:31: warning: 'codecvt_utf8<char32_t, 1114111, 0>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> strCnv;
                              ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/codecvt:199:28: note: 'codecvt_utf8<char32_t, 1114111, 0>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 codecvt_utf8
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
/Users/craig/git-cloud/vhv/verovio/src/vrv.cpp:273:10: warning: 'wstring_convert<std::codecvt_utf8<char32_t, 1114111, 0>, char32_t>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> strCnv;
         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/locale:3621:28: note: 'wstring_convert<std::codecvt_utf8<char32_t, 1114111, 0>, char32_t>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 wstring_convert
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
```

Asking Chat GPT about these warnings, it seems that C++20 (current compiler version for verovio) has a solution:

To resolve the deprecation warnings related to the usage of std::codecvt_utf8 and std::wstring_convert in your C++ code, you need to update the code to use a modern, non-deprecated approach for handling UTF-8 and UTF-32 conversions. The std::codecvt_utf8 and std::wstring_convert classes have been deprecated as of C++17 due to the complexities and limitations they introduce.

# Step 1: Modify UTF-32 to UTF-8 Conversion Function

For converting UTF-32 strings (std::u32string) to UTF-8 (std::string), you can use the char8_t type introduced in C++20. If C++20 is not available, you might need to stick with char and handle UTF-8 encoding explicitly.

Here’s a modified version of the UTF32to8 function using C++11 and later standards without deprecated features:

```cpp
#include <locale>
#include <codecvt>
#include <string>

std::string UTF32to8(const std::u32string &in) {
    std::string out;
    for (auto cp : in) {
        if (cp < 0x80) {  // One byte
            out.push_back(static_cast<char>(cp));
        } else if (cp < 0x800) {  // Two bytes
            out.push_back(static_cast<char>((cp >> 6) | 0xC0));
            out.push_back(static_cast<char>((cp & 0x3F) | 0x80));
        } else if (cp < 0x10000) {  // Three bytes
            out.push_back(static_cast<char>((cp >> 12) | 0xE0));
            out.push_back(static_cast<char>(((cp >> 6) & 0x3F) | 0x80));
            out.push_back(static_cast<char>((cp & 0x3F) | 0x80));
        } else {  // Four bytes
            out.push_back(static_cast<char>((cp >> 18) | 0xF0));
            out.push_back(static_cast<char>(((cp >> 12) & 0x3F) | 0x80));
            out.push_back(static_cast<char>(((cp >> 6) & 0x3F) | 0x80));
            out.push_back(static_cast<char>((cp & 0x3F) | 0x80));
        }
    }
    return out;
}
```

# Step 2: Modify UTF-8 to UTF-32 Conversion Function

Similarly, to convert from UTF-8 to UTF-32:

```cpp
std::u32string UTF8to32(const std::string &in) {
    std::u32string out;
    for (size_t i = 0; i < in.size();) {
        uint32_t cp = in[i] & 0xFF;
        if (cp <= 0x7F) {  // One byte
            out.push_back(cp);
            ++i;
        } else if (cp <= 0xDF) {  // Two bytes
            cp = ((cp & 0x1F) << 6) | (in[i + 1] & 0x3F);
            out.push_back(cp);
            i += 2;
        } else if (cp <= 0xEF) {  // Three bytes
            cp = ((cp & 0x0F) << 12) | ((in[i + 1] & 0x3F) << 6) | (in[i + 2] & 0x3F);
            out.push_back(cp);
            i += 3;
        } else {  // Four bytes
            cp = ((cp & 0x07) << 18) | ((in[i + 1] & 0x3F) << 12) | ((in[i + 2] & 0x3F) << 6) | (in[i + 3] & 0x3F);
            out.push_back(cp);
            i += 4;
        }
    }
    return out;
}
```

# Integration

Replace calls to the deprecated functions with these new functions in your codebase. These implementations directly manage the UTF-8 encoding without using deprecated features. They should be tested thoroughly, especially for edge cases involving characters at the boundaries of UTF-8 byte sequences.

This update will eliminate the deprecation warnings and future-proof your code against upcoming changes in the C++ standards.

**Comment by craigsapp** [comment]
Asking Chat GPT about warnings in toolkit.cpp, here is the response for these warnings:

```
/Users/craig/git-cloud/vhv/verovio/src/toolkit.cpp:377:31: warning: 'codecvt_utf8<char16_t, 1114111, 0>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char16_t>, char16_t> convert;
                              ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/codecvt:199:28: note: 'codecvt_utf8<char16_t, 1114111, 0>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 codecvt_utf8
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
/Users/craig/git-cloud/vhv/verovio/src/toolkit.cpp:377:10: warning: 'wstring_convert<std::codecvt_utf8<char16_t, 1114111, 0>, char16_t>' is deprecated [-Wdeprecated-declarations]
    std::wstring_convert<std::codecvt_utf8<char16_t>, char16_t> convert;
         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/locale:3621:28: note: 'wstring_convert<std::codecvt_utf8<char16_t, 1114111, 0>, char16_t>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 wstring_convert
                           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:837:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__config:810:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
```

Where it discusses a C++17 solution:

To address the deprecation warnings you're seeing in your C++ project, you'll need to replace the usage of std::wstring_convert and std::codecvt_utf8 for UTF-16 strings (char16_t). These classes have been deprecated as of C++17.

Here's a solution using the newer C++17 standard, which supports more modern and potentially safer alternatives for character encoding conversions. However, note that direct replacements within the standard library are limited, so you might need to either implement a custom solution or use a third-party library like boost::locale. For the purpose of this explanation, I'll suggest a method using only the standard library, although it might be less straightforward than using a library designed for such tasks.


# Custom Conversion Function

Since std::codecvt and std::wstring_convert are deprecated and there are no direct replacements in the standard library, you'll need to manually handle the conversion if sticking to the standard library:

```cpp
#include <string>
#include <locale>
#include <codecvt>
#include <vector>

std::string UTF16toUTF8(const std::u16string& input) {
    std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> convert;
    return convert.to_bytes(input);
}

std::u16string UTF8toUTF16(const std::string& input) {
    std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> convert;
    return convert.from_bytes(input);
}
```

# Integration in Your Existing Code


You will replace instances of std::wstring_convert<std::codecvt_utf8<char16_t>, char16_t> with direct calls to these functions.

# Replacement in toolkit.cpp

Find and replace the lines where std::wstring_convert and std::codecvt_utf8<char16_t> are used, replacing them with calls to UTF16toUTF8() or UTF8toUTF16() based on the context in which they are used. This often occurs when you're either saving to or reading from files or other data streams that require text encoding conversions.

# Example Usage

When you need to convert a std::u16string to std::string for UTF-8 encoding, you would do something like:

```cpp
std::u16string data_u16 = u"Example";
std::string data_utf8 = UTF16toUTF8(data_u16);
```

And for converting back:


```cpp
std::string data_utf8 = "Example";
std::u16string data_u16 = UTF8toUTF16(data_utf8);
```

# Caution

If you frequently handle text conversion in your application, consider using a more comprehensive solution like Boost.Locale, which offers extensive support for text processing and character conversions beyond what's provided by the C++ Standard Library. This is especially useful if you need robust error handling and support for a wider range of character encodings.

By moving away from deprecated features, your code not only gains in maintainability but also aligns better with future C++ standards, which emphasize more robust and type-safe alternatives.

**Comment by craigsapp** [comment]
Asking about a C++20 for the toolkit.cpp warnings:

In C++20, handling text encoding transformations like UTF-8 to UTF-16 or vice versa can be managed using standard facilities available in <charconv> for fundamental character conversion operations. However, C++20 does not introduce direct replacements for the full functionality previously provided by std::codecvt and std::wstring_convert. This means that for complete text encoding conversions, additional handling might still be required.

Unfortunately, as of my last update, there are no new standard library features in C++20 that directly replace std::wstring_convert and std::codecvt for arbitrary string conversions. The introduction of char8_t in C++20 formalizes support for UTF-8 character sequences, but utilities for conversion between different Unicode encodings (like UTF-8, UTF-16, UTF-32) in the standard library are still limited.

# Using Third-Party Libraries

Given the lack of direct support in C++20 for these conversions, many developers turn to third-party libraries for robust support. Libraries like Boost.Locale or ICU (International Components for Unicode) are often used in environments where comprehensive text processing and Unicode support are required.

# Manual Conversion Example in C++20

If sticking to standard C++ and handling basic scenarios without external libraries, you would typically need to manually handle UTF-16 to UTF-8 conversions as shown earlier. For a more "modern" approach utilizing C++20 features, you can still use facilities like std::u8string for UTF-8 and continue handling conversions manually:

```cpp
#include <string>
#include <vector>

std::u8string UTF16toUTF8(const std::u16string& input) {
    std::u8string output;
    // Placeholder for manual conversion logic
    // Real conversion logic here should handle actual UTF-16 to UTF-8 conversion
    for (char16_t c : input) {
        if (c < 0x80) {  // Handle basic ASCII conversion
            output.push_back(static_cast<char8_t>(c));
        } else {
            // Extend this block to handle non-ASCII characters
        }
    }
    return output;
}

std::u16string UTF8toUTF16(const std::u8string& input) {
    std::u16string output;
    // Placeholder for manual conversion logic
    // Real conversion logic here should handle actual UTF-8 to UTF-16 conversion
    for (char8_t c : input) {
        output.push_back(static_cast<char16_t>(c));  // Naive conversion for ASCII only
    }
    return output;
}
```


In C++20, handling text encoding transformations like UTF-8 to UTF-16 or vice versa can be managed using standard facilities available in <charconv> for fundamental character conversion operations. However, C++20 does not introduce direct replacements for the full functionality previously provided by std::codecvt and std::wstring_convert. This means that for complete text encoding conversions, additional handling might still be required.

Unfortunately, as of my last update, there are no new standard library features in C++20 that directly replace std::wstring_convert and std::codecvt for arbitrary string conversions. The introduction of char8_t in C++20 formalizes support for UTF-8 character sequences, but utilities for conversion between different Unicode encodings (like UTF-8, UTF-16, UTF-32) in the standard library are still limited.

Using Third-Party Libraries
Given the lack of direct support in C++20 for these conversions, many developers turn to third-party libraries for robust support. Libraries like Boost.Locale or ICU (International Components for Unicode) are often used in environments where comprehensive text processing and Unicode support are required.

Manual Conversion Example in C++20
If sticking to standard C++ and handling basic scenarios without external libraries, you would typically need to manually handle UTF-16 to UTF-8 conversions as shown earlier. For a more "modern" approach utilizing C++20 features, you can still use facilities like std::u8string for UTF-8 and continue handling conversions manually:

```cpp
#include <string>
#include <vector>

std::u8string UTF16toUTF8(const std::u16string& input) {
    std::u8string output;
    // Placeholder for manual conversion logic
    // Real conversion logic here should handle actual UTF-16 to UTF-8 conversion
    for (char16_t c : input) {
        if (c < 0x80) {  // Handle basic ASCII conversion
            output.push_back(static_cast<char8_t>(c));
        } else {
            // Extend this block to handle non-ASCII characters
        }
    }
    return output;
}

std::u16string UTF8toUTF16(const std::u8string& input) {
    std::u16string output;
    // Placeholder for manual conversion logic
    // Real conversion logic here should handle actual UTF-8 to UTF-16 conversion
    for (char8_t c : input) {
        output.push_back(static_cast<char16_t>(c));  // Naive conversion for ASCII only
    }
    return output;
}
```

# Recommendations

1 For simple ASCII data: The above manual methods can be sufficient.
2 For full Unicode support: Consider using a library like Boost.Locale or even smaller, more focused libraries that are intended for Unicode handling.
3 For projects where dependencies are not a concern: Libraries like ICU provide extensive support for all forms of Unicode handling, including conversions across different encodings, locale-specific operations, and more.

In practice, handling complex Unicode data in C++ often still involves leveraging external libraries, especially in applications needing comprehensive internationalization support.

**Comment by lpugin** [comment]
@craigsapp [this](https://github.com/rism-digital/verovio/discussions/3651#discussion-6531060) looks like the most appropriate solution for UTF8 to UTF32. I will give a try.

**Reply by lpugin** [reply]
See https://github.com/rism-digital/verovio/commit/2e7290c1f497ed6bbc73ea2d07a909fd15ae7047 and dir-010 and dir-011 in the test suite.

**Comment by lpugin** [comment]
Fixed with https://github.com/rism-digital/verovio/commit/2e7290c1f497ed6bbc73ea2d07a909fd15ae7047
